File: ui/src/lib/monitoring.ts
export type FindingSeverity = "critical" | "high" | "medium" | "low" | "info";
export type FindingStatus = "open" | "acknowledged" | "accepted_risk" | "resolved";
export type EvidenceState = "exact" | "bounded" | "indeterminate";

export type FindingEvidence = {
  state: EvidenceState;
  summary?: string | null;
  refs?: Array<Record<string, unknown> | string> | null;
  limitations?: string[] | null;
};

export type Finding = {
  id: string;
  project_id: string;
  source_id?: string | null;
  policy_id: string;
  policy_version: number;
  title: string;
  description?: string | null;
  severity: FindingSeverity;
  status: FindingStatus;
  resource_identity_key?: string | null;
  resource_type?: string | null;
  provider?: string | null;
  resource_name?: string | null;
  first_seen_at: string;
  last_seen_at: string;
  resolved_at?: string | null;
  accepted_risk_expires_at?: string | null;
  assignee_user_id?: string | null;
  latest_run_id?: string | null;
  latest_comparison_id?: string | null;
  evidence: FindingEvidence;
  occurrence_count: number;
  revision: number;
  created_at: string;
  updated_at: string;
};

export type FindingSummary = Record<FindingStatus, number> & { total: number };

export type FindingPolicy = {
  id: string;
  version: number;
  title: string;
  description?: string | null;
  severity: FindingSeverity;
  category?: string | null;
  enabled: boolean;
  evidence_requirements?: string[] | Record<string, unknown> | null;
};

export type FindingOccurrence = {
  id: number;
  run_id: string;
  comparison_id?: string | null;
  policy_id: string;
  policy_version: number;
  evidence_state: EvidenceState;
  evidence?: FindingEvidence | Record<string, unknown> | null;
  observed_at: string;
};

export type FindingActivity = {
  id: number;
  ts: string;
  action: string;
  actor_user_id?: string | null;
  actor_token_id?: string | null;
  metadata?: Record<string, unknown> | null;
};

export type SourceCoverage = {
  state: "complete" | "partial" | "unknown";
  reasons?: string[] | null;
};

export type SourceFreshness = {
  state: "fresh" | "stale" | "unknown" | "disabled";
  age_seconds?: number | null;
  expected_interval_seconds?: number | null;
  stale_after_seconds?: number | null;
};

export type SourceHealth = "healthy" | "stale" | "degraded" | "never_collected" | "disabled";

export type MonitoringSource = {
  id: string;
  project_id: string;
  source_key: string;
  display_name: string;
  provider: string;
  assessed_identity?: string | null;
  target_scope?: Record<string, unknown> | null;
  enabled: boolean;
  expected_interval_seconds?: number | null;
  last_run_id?: string | null;
  last_success_at?: string | null;
  last_failure_at?: string | null;
  last_comparison_id?: string | null;
  collector_version?: string | null;
  coverage: SourceCoverage;
  freshness: SourceFreshness;
  health_status: SourceHealth;
  health_reasons?: string[] | null;
  created_at: string;
  updated_at: string;
};

export type EffectiveAccessPrincipal = {
  id?: string | number | null;
  display_name?: string | null;
  principal_type?: string | null;
  kind?: string | null;
  provider_principal_id?: string | null;
  principal_key?: string | null;
  login_name?: string | null;
  email?: string | null;
  resolution?: string | null;
  direct_decision?: string | null;
  effective_decision?: string | null;
  decision?: string | null;
  capabilities?: string[] | Record<string, unknown> | null;
  allow_rights?: string[] | null;
  deny_rights?: string[] | null;
  limitations?: string[] | null;
  explanation?: string | null;
  paths?: string[] | null;
  [key: string]: unknown;
};

export type EffectiveAccessAnalysis = {
  resource?: Record<string, unknown> | null;
  analysis_state: "computed" | "bounded" | "indeterminate";
  decision: "allowed" | "denied" | "mixed" | "unknown";
  capabilities?: string[] | Record<string, unknown> | null;
  principals: { items: EffectiveAccessPrincipal[]; next_cursor?: string | null };
  evidence_planes?: Record<string, unknown> | null;
  limitations?: string[] | null;
  provenance?: Record<string, unknown> | null;
};

export function humanizeMonitoringValue(value: string | null | undefined): string {
  if (!value) return "Not recorded";
  return value
    .trim()
    .replaceAll("-", "_")
    .split("_")
    .filter(Boolean)
    .map((part) => part.charAt(0).toUpperCase() + part.slice(1))
    .join(" ");
}

export function findingSeverityRank(severity: FindingSeverity): number {
  return ({ critical: 5, high: 4, medium: 3, low: 2, info: 1 })[severity];
}

export function findingTone(severity: FindingSeverity): "critical" | "high" | "medium" | "low" | "info" {
  return severity;
}

export function canManageFindings(role: string | null | undefined): boolean {
  return role === "operator" || role === "admin";
}

export function canManageSources(role: string | null | undefined): boolean {
  return role === "admin";
}

export function evidenceTrustCopy(state: EvidenceState): string {
  if (state === "exact") return "Exact within the declared collection scope.";
  if (state === "bounded") return "Based on bounded observations; review limitations before acting.";
  return "Collection evidence is insufficient for a definitive conclusion.";
}

export function formatMonitoringTimestamp(value: string | null | undefined): string {
  if (!value) return "Not recorded";
  const date = new Date(value);
  return Number.isNaN(date.getTime()) ? value : date.toLocaleString();
}

export function formatDuration(seconds: number | null | undefined): string {
  if (seconds == null || !Number.isFinite(seconds) || seconds < 0) return "Unknown";
  if (seconds < 60) return `${Math.round(seconds)}s`;
  if (seconds < 3600) return `${Math.round(seconds / 60)}m`;
  if (seconds < 86400) return `${Math.round(seconds / 3600)}h`;
  return `${Math.round(seconds / 86400)}d`;
}

File: api/app/services/monitoring.py
from __future__ import annotations

from datetime import UTC, datetime
from typing import Any

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.comparison_contract import COMPARISON_ALGORITHM_VERSION, comparison_algorithm_is_current
from app.locking import lock_monitoring_source
from app.models import CollectionSource, Finding, RunComparison

BUILT_IN_FINDING_POLICIES: tuple[dict[str, Any], ...] = (
    {
        "id": "sharepoint.anonymous_access",
        "version": 1,
        "title": "Anonymous SharePoint access",
        "description": "A SharePoint resource has explicit evidence of an anonymous sharing link.",
        "severity": "critical",
        "category": "exposure",
        "enabled": True,
        "mode": "state",
        "evidence_requirements": ["explicit anonymous link scope"],
    },
    {
        "id": "sharepoint.broad_internal_access",
        "version": 1,
        "title": "Organization-wide SharePoint access",
        "description": "A SharePoint resource has explicit organization-wide sharing evidence.",
        "severity": "medium",
        "category": "exposure",
        "enabled": True,
        "mode": "state",
        "evidence_requirements": ["explicit organization link scope"],
    },
    {
        "id": "smb.write_observed",
        "version": 1,
        "title": "SMB write capability observed",
        "description": "One or more bounded, non-mutating SMB capability probes indicate write access.",
        "severity": "high",
        "category": "access",
        "enabled": True,
        "mode": "state",
        "evidence_requirements": ["allowed or mixed write capability observation", "finalized probe evidence"],
    },
    {
        "id": "resource.appeared",
        "version": 1,
        "title": "Resource appeared",
        "description": "A resource appeared between structurally comparable collection runs.",
        "severity": "info",
        "category": "change",
        "enabled": True,
        "mode": "event",
        "evidence_requirements": ["structurally comparable runs"],
    },
    {
        "id": "resource.disappeared",
        "version": 1,
        "title": "Resource disappeared",
        "description": "A resource disappeared between structurally comparable collection runs.",
        "severity": "low",
        "category": "change",
        "enabled": True,
        "mode": "event",
        "evidence_requirements": ["authoritative structurally comparable runs"],
    },
    {
        "id": "permission.evidence_changed",
        "version": 1,
        "title": "Permission evidence changed",
        "description": "Comparable direct-permission or bounded capability evidence changed.",
        "severity": "high",
        "category": "access_change",
        "enabled": True,
        "mode": "event",
        "evidence_requirements": ["comparable permission or capability evidence"],
    },
    {
        "id": "comparison.indeterminate",
        "version": 1,
        "title": "Change is indeterminate",
        "description": "Collection coverage or identity continuity does not support a definitive change conclusion.",
        "severity": "low",
        "category": "coverage",
        "enabled": True,
        "mode": "event",
        "evidence_requirements": ["materialized comparison with an indeterminate result"],
    },
)


class AutomaticSourceDisabledError(RuntimeError):
    pass

POLICY_BY_ID = {policy["id"]: policy for policy in BUILT_IN_FINDING_POLICIES}
FINDING_STATUSES = frozenset({"open", "acknowledged", "accepted_risk", "resolved"})
FINDING_SEVERITIES = frozenset({"critical", "high", "medium", "low", "info"})
SOURCE_HEALTH_STATES = frozenset({"healthy", "stale", "degraded", "never_collected", "disabled"})


def utc_datetime(value: datetime | None) -> datetime | None:
    if value is None:
        return None
    if value.tzinfo is None:
        return value.replace(tzinfo=UTC)
    return value.astimezone(UTC)


def source_health(source: CollectionSource, *, now: datetime | None = None) -> dict[str, Any]:
    current_time = utc_datetime(now) or datetime.now(tz=UTC)
    success_at = utc_datetime(source.last_success_at)
    failure_at = utc_datetime(source.last_failure_at)
    expected = source.expected_interval_seconds
    age_seconds = max(0, int((current_time - success_at).total_seconds())) if success_at else None
    stale_after = max(900, expected * 2) if expected is not None else None

    coverage = dict(source.coverage or {})
    coverage_state = str(coverage.get("state") or "unknown").lower()
    coverage_reasons = [str(reason) for reason in coverage.get("reasons", []) if str(reason).strip()]
    findings_coverage = (
        coverage.get("monitoring_findings") if isinstance(coverage.get("monitoring_findings"), dict) else {}
    )
    baseline_coverage = (
        coverage.get("automatic_baseline") if isinstance(coverage.get("automatic_baseline"), dict) else {}
    )
    findings_state = str(findings_coverage.get("state") or "unknown").lower()
    baseline_state = str(baseline_coverage.get("state") or "unknown").lower()
    baseline_findings_state = str(baseline_coverage.get("findings_evaluation_state") or "unknown").lower()
    baseline_algorithm_current = comparison_algorithm_is_current(baseline_coverage.get("algorithm_version"))
    monitoring_complete = (
        findings_state == "complete"
        and baseline_state == "established"
        and baseline_findings_state == "complete"
        and baseline_algorithm_current
    )
    reasons: list[str] = []
    if not source.enabled:
        health_status = "disabled"
        freshness_state = "disabled"
        reasons.append("Monitoring is disabled for this source.")
    elif success_at is None:
        health_status = "never_collected"
        freshness_state = "unknown"
        reasons.append("No successful collection has been registered.")
    elif stale_after is not None and age_seconds is not None and age_seconds > stale_after:
        health_status = "stale"
        freshness_state = "stale"
        reasons.append("The last successful collection is older than the configured freshness window.")
    elif failure_at is not None and failure_at > success_at:
        health_status = "degraded"
        freshness_state = "fresh" if stale_after is None or age_seconds <= stale_after else "stale"
        reasons.append("The most recent collection attempt failed after the last success.")
    elif coverage_state != "complete" or not monitoring_complete:
        health_status = "degraded"
        freshness_state = "fresh"
        reasons.extend(coverage_reasons or ["The latest collection reported partial or unknown coverage."])
        if baseline_state == "established" and not baseline_algorithm_current:
            reasons.append("Automatic comparison coverage uses an obsolete or unknown algorithm.")
    else:
        health_status = "healthy"
        freshness_state = "fresh"

    return {
        "health_status": health_status,
        "health_reasons": list(dict.fromkeys(reasons)),
        "coverage": {
            **coverage,
            "automatic_baseline": {
                **baseline_coverage,
                "algorithm_current": baseline_algorithm_current,
            },
            "state": coverage_state if coverage_state in {"complete", "partial", "unknown"} else "unknown",
            "reasons": coverage_reasons,
        },
        "freshness": {
            "state": freshness_state,
            "age_seconds": age_seconds,
            "expected_interval_seconds": expected,
            "stale_after_seconds": stale_after,
        },
    }


def source_payload(source: CollectionSource, *, now: datetime | None = None) -> dict[str, Any]:
    health = source_health(source, now=now)
    return {
        "id": str(source.id),
        "project_id": str(source.project_id),
        "source_key": source.source_key,
        "display_name": source.display_name,
        "provider": source.provider,
        "assessed_identity": source.assessed_identity,
        "target_scope": source.target_scope or {},
        "enabled": source.enabled,
        "expected_interval_seconds": source.expected_interval_seconds,
        "last_run_id": str(source.last_run_id) if source.last_run_id else None,
        "last_success_at": source.last_success_at.isoformat() if source.last_success_at else None,
        "last_failure_at": source.last_failure_at.isoformat() if source.last_failure_at else None,
        "last_comparison_id": str(source.last_comparison_id) if source.last_comparison_id else None,
        "collector_version": source.collector_version,
        **health,
        "created_at": source.created_at.isoformat() if source.created_at else None,
        "updated_at": source.updated_at.isoformat() if source.updated_at else None,
    }


def publish_automatic_baseline_recovery(
    db: Session,
    comparison: RunComparison,
    *,
    findings_only: bool,
    require_enabled: bool = False,
) -> bool:
    """Publish recovery only while this comparison is the source's current baseline."""

    if (
        comparison.trigger != "automatic"
        or comparison.source_id is None
        or not comparison_algorithm_is_current(comparison.algorithm_version)
    ):
        return False
    candidate = db.get(CollectionSource, comparison.source_id)
    if candidate is None:
        return False
    lock_monitoring_source(db, candidate.source_key)
    source = db.execute(
        select(CollectionSource)
        .where(
            CollectionSource.id == comparison.source_id,
            CollectionSource.project_id == comparison.project_id,
        )
        .with_for_update()
        .execution_options(populate_existing=True)
    ).scalar_one_or_none()
    if source is None:
        return False
    if require_enabled and not source.enabled:
        raise AutomaticSourceDisabledError("automatic comparison source is disabled")
    coverage = dict(source.coverage or {})
    baseline = (
        dict(coverage.get("automatic_baseline"))
        if isinstance(coverage.get("automatic_baseline"), dict)
        else {}
    )
    if str(baseline.get("comparison_id") or "") != str(comparison.id):
        return False
    if findings_only:
        baseline.update(
            {
                "state": "established",
                "algorithm_version": COMPARISON_ALGORITHM_VERSION,
                "findings_evaluation_state": "queued",
                "findings_next_retry_at": None,
            }
        )
    else:
        baseline.update(
            {
                "state": "queued",
                "algorithm_version": COMPARISON_ALGORITHM_VERSION,
                "findings_evaluation_state": "queued",
                "findings_next_retry_at": None,
                "next_retry_at": None,
            }
        )
    baseline.pop("error_code", None)
    baseline.pop("reason", None)
    coverage["automatic_baseline"] = baseline
    coverage["state"] = "partial"
    reasons = [str(reason) for reason in coverage.get("reasons", []) if str(reason).strip()]
    reasons.append(
        "Automatic comparison finding evaluation is pending or retrying."
        if findings_only
        else "Automatic comparison recovery is queued."
    )
    coverage["reasons"] = list(dict.fromkeys(reasons))
    source.coverage = coverage
    source.updated_at = datetime.now(tz=UTC)
    return True


def finding_payload(finding: Finding, *, include_evidence: bool = True) -> dict[str, Any]:
    evidence = finding.evidence or {}
    public_evidence = evidence if include_evidence else {"state": str(evidence.get("state") or "indeterminate")}
    return {
        "id": str(finding.id),
        "project_id": str(finding.project_id),
        "source_id": str(finding.source_id) if finding.source_id else None,
        "policy_id": finding.policy_id,
        "policy_version": finding.policy_version,
        "title": finding.title,
        "description": finding.description,
        "severity": finding.severity,
        "status": finding.status,
        "resource_identity_key": finding.resource_identity_key,
        "resource_type": finding.resource_type,
        "provider": finding.provider,
        "resource_name": finding.resource_name,
        "first_seen_at": finding.first_seen_at.isoformat() if finding.first_seen_at else None,
        "last_seen_at": finding.last_seen_at.isoformat() if finding.last_seen_at else None,
        "resolved_at": finding.resolved_at.isoformat() if finding.resolved_at else None,
        "accepted_risk_expires_at": (
            finding.accepted_risk_expires_at.isoformat() if finding.accepted_risk_expires_at else None
        ),
        "assignee_user_id": str(finding.assignee_user_id) if finding.assignee_user_id else None,
        "latest_run_id": str(finding.latest_run_id) if finding.latest_run_id else None,
        "latest_comparison_id": str(finding.latest_comparison_id) if finding.latest_comparison_id else None,
        "evidence": public_evidence,
        "occurrence_count": finding.occurrence_count,
        "revision": finding.revision,
        "created_at": finding.created_at.isoformat() if finding.created_at else None,
        "updated_at": finding.updated_at.isoformat() if finding.updated_at else None,
    }
